wallet/walletrpc.c

```diff
@@ -191,27 +223,106 @@ static struct command_result *json_prepare_tx(struct command *cmd,
 	}
 
 	maxheight = minconf_to_maxheight(*minconf, cmd->ld);
+
+	/* *withdraw* command or old *txprepare* command.
+	 * Support only one output. */
+	if (destination) {
+		outputs = tal_arr(tmpctx, struct bitcoin_tx_output *, 1);
+		outputs[0] = tal(outputs, struct bitcoin_tx_output);
+		outputs[0]->script = tal_steal(outputs[0],
+					       cast_const(u8 *, destination));
+		outputs[0]->amount = (*utx)->wtx->amount;
+		out_len = tal_count(outputs[0]->script);
+
+		goto create_tx;
+	}
+
+	if (outputstok->size == 0)
+		return command_fail(cmd, JSONRPC2_INVALID_PARAMS, "Empty outputs");
+
+	outputs = tal_arr(tmpctx, struct bitcoin_tx_output *, outputstok->size);
+	out_len = 0;
+	(*utx)->wtx->all_funds = false;
+	json_for_each_arr(i, t, outputstok) {
+		struct amount_sat *amount;
+		const u8 *destination;
+		enum address_parse_result res;
+
+		/* output format: {destination: amount} */
+		if (t->type != JSMN_OBJECT)
+			return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+					    "The output format must be "
+					    "{destination: amount}");;
+
+		res = json_tok_address_scriptpubkey(cmd,
+						    get_chainparams(cmd->ld),
+						    buffer, &t[1],
+						    &destination);
+		if (res == ADDRESS_PARSE_UNRECOGNIZED)
+			return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+					    "Could not parse destination address");
+		else if (res == ADDRESS_PARSE_WRONG_NETWORK)
+			return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+					    "Destination address is not on network %s",
+					    get_chainparams(cmd->ld)->network_name);
+
+		amount = tal(tmpctx, struct amount_sat);
+		if (!json_to_sat_or_all(buffer, &t[2], amount))
+			return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+					    "'%.*s' is a invalid satoshi amount",
+					    t[2].end - t[2].start, buffer + t[2].start);
+
+		out_len += tal_count(destination);
+		outputs[i] = tal(outputs, struct bitcoin_tx_output);
+		outputs[i]->amount = *amount;
+		outputs[i]->script = tal_steal(outputs[i],
+					       cast_const(u8 *, destination));
+
+		/* In fact, the maximum amount of bitcoin satoshi is 2.1e15.
+		 * It can't be equal to/bigger than 2^64.
+		 * On the hand, the maximum amount of litoshi is 8.4e15,
+		 * which also can't overflow. */
+		/* This means this destination need "all" satoshi we have. */
+		if (amount_sat_eq(*amount, AMOUNT_SAT(-1ULL))) {
+			if (outputstok->size > 1)
+				return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+						    "outputs[%zi]: this destination wants"
+						    " all satoshi. The amount of outputs"
+						    " can't be more than 1. ", i);
+			(*utx)->wtx->all_funds = true;
+			break;
+		}
+
+		if (i == 0)
+			(*utx)->wtx->amount = AMOUNT_SAT(0);
```

Comment:
we should do this before the loop, right after we initialize `all_funds = false`

Reply:
Yes, thank you. Corrected.

Reply:
Here shouldn't be changed:
At the beginning, we initial the `(*utx)->wtx->amount` as `AMOUNT_SAT(-1ULL)`.
This is the max permissible for `wallet_select_all`.
```
	if (amount_sat_eq(*amount, AMOUNT_SAT(-1ULL))) {
		if (outputstok->size > 1)
			return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
					    "outputs[%zi]: this destination wants"
					    " all satoshi. The count of outputs"
					    " can't be more than 1. ", i);
		(*utx)->wtx->all_funds = true;
		break;
	}

	if (i == 0)
		(*utx)->wtx->amount = AMOUNT_SAT(0);
```

After handling the `all` case, we can initial `(*utx)->wtx->amount` again.

Reply:
then we're in a split mode, because you've initialized `all_funds` to be false, but left the wtx->amount pegged at what basically signifies `all_funds`. this is not ideal.

the wtx amount should be set to zero at the loop start along with the `all_funds` flag. in the case that the `*amount` is the max, then set (utx)->wtx->amount to be *amount before breaking. this keeps the state fields of the wtx in sync.

Reply:
Yes, I got your idea. It's better :+1: